File: PAT_Basic_Level_by_java/1110 区块反转/1110 区块反转.java
/*
 * 1110 区块反转
 * 
 * 题目描述：
 *     给定一个单链表L和一个正整数K，将链表按每K个节点为一个区块
 *     进行划分，然后反转区块的顺序（注意：区块内节点顺序保持不变）。
 *     输出反转后的链表。
 *     例如：链表1→2→3→4→5→6→7，K=3，
 *           区块划分为[1,2,3] [4,5,6] [7]，
 *           反转区块顺序后为[7] [4,5,6] [1,2,3]。
 * 
 * 实现原理：
 *     解题思路：
 *         1. 使用HashMap以节点地址为键存储所有节点。
 *         2. 从头节点开始遍历链表，将节点按顺序存入List。
 *         3. 将List按每K个节点切分为区块（最后一个区块可能不足K个）。
 *         4. 反转区块的顺序（整个区块为单位反转，区块内节点顺序不变）。
 *         5. 遍历反转后的结果列表，更新每个节点的next指针指向下一节点地址。
 *     时间复杂度：O(N)，空间复杂度：O(N)。
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Main {
    // 链表节点类，包含地址、数据和下一节点地址
    static class Node {
        String address;
        int data;
        String next;
        Node(String addr, int d, String n) {
            address = addr;
            data = d;
            next = n;
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstLine = br.readLine().split(" ");
        String head = firstLine[0];          // 链表头节点地址
        int n = Integer.parseInt(firstLine[1]);  // 节点总数
        int k = Integer.parseInt(firstLine[2]);  // 区块大小
        
        // 用HashMap存储所有节点，地址作为键
        Map<String, Node> nodes = new HashMap<>();
        for (int i = 0; i < n; i++) {
            String[] parts = br.readLine().split(" ");
            nodes.put(parts[0], new Node(parts[0], Integer.parseInt(parts[1]), parts[2]));
        }
        
        // 从头节点开始遍历，按顺序构建链表
        List<Node> list = new ArrayList<>();
        String curr = head;
        while (!curr.equals("-1")) {
            Node node = nodes.get(curr);
            list.add(node);
            curr = node.next;  // 移动到下一个节点
        }
        
        // 将链表分成区块，每K个节点为一个区块（最后一个区块可能不足K个）
        List<List<Node>> blocks = new ArrayList<>();
        for (int i = 0; i < list.size(); i += k) {
            int end = Math.min(i + k, list.size());
            blocks.add(new ArrayList<>(list.subList(i, end)));
        }
        
        // 反转区块的顺序（区块整体反转，而非区块内节点反转）
        List<Node> result = new ArrayList<>();
        for (int i = blocks.size() - 1; i >= 0; i--) {
            result.addAll(blocks.get(i));
        }
        
        // 输出反转后的链表，更新每个节点的next指针
        for (int i = 0; i < result.size(); i++) {
            Node node = result.get(i);
            String nextAddr = (i == result.size() - 1) ? "-1" : result.get(i + 1).address;
            System.out.println(node.address + " " + node.data + " " + nextAddr);
        }
    }
}

File: PAT_Basic_Level_by_cpp/1110 区块反转/1110 区块反转.cpp
/*
 * 1110 区块反转
 *
 * 【实现原理】
 * 题目给定一个单链表和一个整数 K，要求将链表每 K 个结点分为一块，
 * 反转"块"的顺序（即块与块之间交换位置，块内结点保持原有顺序），
 * 最后输出反转后的链表。
 *
 * 【算法思路】
 * 1. 使用固定大小数组（100000）以地址为下标存储所有结点。
 * 2. 从头结点开始遍历链表，将结点地址按原始顺序收集到 order 数组中。
 * 3. 将 order 按每 K 个结点一组（最后一块可能不足 K 个）切分成多个块。
 * 4. 反转各块的顺序（即块整体位置反转）。
 * 5. 将反转后的块按顺序重新拼接，得到新的结点序列。
 * 6. 按新序列顺序输出结点（地址、数据、下一结点地址）。
 *
 * 【复杂度分析】
 * - 时间复杂度：O(N)，遍历链表一次，分块和反转操作均为 O(N)。
 * - 空间复杂度：O(N + 100000)，存储结点信息和序列。
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdio>

using namespace std;

struct Node {
    int addr, data, next;  // 结点地址、数据、下一结点地址
};

int main() {
    int head, n, k;
    if (!(cin >> head >> n >> k)) return 0;

    vector<Node> nodes(100000);                  // 以地址为下标存储结点
    for (int i = 0; i < n; ++i) {
        int a, d, ne; cin >> a >> d >> ne; nodes[a] = {a, d, ne};
    }

    vector<int> order;
    // 遍历链表，按原始顺序收集结点地址
    for (int p = head; p != -1; p = nodes[p].next) order.push_back(p);

    // 将 order 按每 K 个结点分为一块
    vector<vector<int>> blocks;
    for (size_t i = 0; i < order.size(); i += k)
        blocks.emplace_back(order.begin() + i, order.begin() + i + min((size_t)k, order.size() - i));
    reverse(blocks.begin(), blocks.end());       // 反转块顺序

    vector<int> res;
    // 将反转后的块重新拼接为序列
    for (auto& b : blocks) for (int x : b) res.push_back(x);

    // 输出反转后的链表
    for (size_t i = 0; i < res.size(); ++i) {
        int addr = res[i];
        int nextAddr = (i + 1 < res.size()) ? res[i + 1] : -1;
        if (nextAddr == -1) printf("%05d %d -1\n", addr, nodes[addr].data);
        else printf("%05d %d %05d\n", addr, nodes[addr].data, nextAddr);
    }
    return 0;
}